### ISSUE TITLE
group_by:+when: skips hosts, depending on inventory order

### ISSUE THREAD

**Issue by mark-casey** [ansible/ansible#7199]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.5.4
##### Environment:

Ubuntu 12.04 x64 running Ansible, managing 13.10 x64 hosts
##### Summary:
- Attempting to use 'group_by:' with 'when:'.
- Two hosts; one in group _**dbnodes**_, one in group _**webnodes**_, both also a member of ec2_hosts.
- One host satisfies the when clause, one does not.
- Flipping group membership doesn't flip which host is grouped...just skips the task altogether.
##### Steps To Reproduce:

See gist: https://gist.github.com/mark-casey/11386608
One provided inventory works for me, one doesn't.
##### Expected Results:

The host which is a member of 'dbnodes' should be added to the 'Debian' group, for either of the two inventory files provided.
##### Actual Results:

One of the inventory files sees no hosts added to the 'Debian' group.

Also interesting, either of the two inventories will work if you entirely remove the ec2_hosts group and change site.yml's play to use this host selector instead:

```
  hosts: dbnodes:webnodes
```

**Comment by mpdehaan** [comment]
"Flipping group membership doesn't flip which host is grouped...just skips the task altogether."

This is exactly what when does, when something is not in a "when" clause, that statement is not run.

It sounds like you have some more basic questions about this rather than a bug report, how about stopping by the mailing list?

**Comment by mark-casey** [comment]
The host that satisfies the when clause is present in both inventories. The task just behaves differently based on whether it first evaluates the host that matches or the host that does not.

Were you able to try the sample case(s)?

**Comment by jimi-c** [comment]
@mark-casey I can confirm I see the same behavior, and also I see that if you flip the .79/.71 hosts in the [ec2_hosts] group it works as expected. Re-opening for further investigation.

**Comment by mark-casey** [comment]
Thank you guys for taking another look. (I fully expected it might need some clarification upon hearing friend1 explain it to friend2 [and I quote]: "It's a slot machine written in Ansible. Sometimes you win and the service comes up, and sometimes the house/ec2 DHCP server cheats you.")

**Comment by sivel** [comment]
I may be wrong here, but this is probably caused by group_by being a host loop bypass plugin.  So it does not run for every host, basically just the first fork which could be an arbitrary host.

**Comment by mpdehaan** [comment]
This is correct.   If so we could just make an error that group_by can't be used with when.

**Comment by mark-casey** [comment]
@mpdehaan I have to admit that in retrospect this feels like one of those things for which there would always be a better way, but that feels like an awfully wide brush. Is the combination bad enough that it would've ended up forbidden even if it were working as expected?

I've backed out my related changes and will seek a better way, but maybe the logic could be pushed further into the one execution of the module you do get? Perhaps a parameter with semantics similar to a play's 'hosts:' selector? and then _also_ forbid group_by+when?

```
group_by: key=foo hosts=dbnodes!webnodes
```

I guess that begs the question of what happens when you list hosts that aren't in the play though.

edit: or maybe   "key=foo no_consider=badgroup"

**Comment by semifocused** [comment]
Hi guys - I don't know where you've left this one, but I did just spend a few hours isolating an issue in my playbooks to the fact that "group_by:" currently behaves unpredictably with "when:"

If there's no plan to resolve, it would be helpful to update the docs.

I was using when: to rule out instances for which a part of the group_by key would be undefined.  But that led to a single host being skipped and no groups being created.  Because I'm not familiar with the internals of the "host loop bypass plugin", it took me a while to diagnose this and track down this thread.

I was able to work around it be creating an initial group that contains those hosts for which all the segments of the key are defined.  Then I run a new play with only those hosts and no "when: with the group_by.

Now its working well, great work on the module.

**Comment by yoshida-mediba** [comment]
:star:

**Comment by jimi-c** [comment]
Looking back at this issue, to clarify, the problem is that the `when` statement is using an inventory variable (in the original example, group_names). As @sivel pointed out, `group_by` is a host loop bypass action, meaning it does not operate on the full list of hosts currently in the play. Essentially, the first host in the list gets the task so only the inventory variables from that host are used, which is why the ordering here is important.

The architecture on the current devel post v2 (ansible 2.0) has not changed much in this regard, so the problem persists. The best course of action may be some documentation notes/warnings on either the group_by module (and any other host-loop bypass actions, such as debug, etc.) indicating that using `when` statements that rely on inventory variables may have surprising results based on the inventory ordering, or even a special section in the conditionals documentation.

**Comment by jimi-c** [comment]
Hi @mark-casey, sorry for the late follow-up. It appears that this has been resolved since the 2.0 release, we just never went back and confirmed it. Using anything from v2.0.0.2-1 on (which is merely the furthest I went back), I see the correct output using both inventories:

<pre>
$ ansible-playbook -vv -i inv.bad test.yml
Using /etc/ansible/ansible.cfg as config file
1 plays in test.yml
PLAY [Custom re-group] *********************************************************
TASK [setup] *******************************************************************
ok: [172.31.29.71]
ok: [172.31.22.79]
TASK [debug] *******************************************************************
ok: [172.31.29.71] => {
    "group_names": [
        "ec2_group", 
        "dbnodes"
    ]
}
ok: [172.31.22.79] => {
    "group_names": [
        "webnodes", 
        "ec2_group"
    ]
}
TASK [Add a dbnode host to group] **********************************************
skipping: [172.31.22.79] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [172.31.29.71] => {"add_group": "RedHat", "changed": false}
TASK [debug] *******************************************************************
ok: [172.31.22.79] => {
    "group_names": [
        "webnodes", 
        "ec2_group"
    ]
}
ok: [172.31.29.71] => {
    "group_names": [
        "ec2_group", 
        "dbnodes", 
        "RedHat"
    ]
}
PLAY RECAP *********************************************************************
172.31.22.79               : ok=3    changed=0    unreachable=0    failed=0   
172.31.29.71               : ok=4    changed=0    unreachable=0    failed=0   
[jamesc@jimi 7199]$ ansible-playbook -vv -i inv.good test.yml
Using /etc/ansible/ansible.cfg as config file
1 plays in test.yml
PLAY [Custom re-group] *********************************************************
TASK [setup] *******************************************************************
ok: [172.31.22.79]
ok: [172.31.29.71]
TASK [debug] *******************************************************************
ok: [172.31.22.79] => {
    "group_names": [
        "ec2_group", 
        "dbnodes"
    ]
}
ok: [172.31.29.71] => {
    "group_names": [
        "webnodes", 
        "ec2_group"
    ]
}
TASK [Add a dbnode host to group] **********************************************
skipping: [172.31.29.71] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
ok: [172.31.22.79] => {"add_group": "RedHat", "changed": false}
TASK [debug] *******************************************************************
ok: [172.31.22.79] => {
    "group_names": [
        "ec2_group", 
        "dbnodes", 
        "RedHat"
    ]
}
ok: [172.31.29.71] => {
    "group_names": [
        "webnodes", 
        "ec2_group"
    ]
}
PLAY RECAP *********************************************************************
172.31.22.79               : ok=4    changed=0    unreachable=0    failed=0   
172.31.29.71               : ok=3    changed=0    unreachable=0    failed=0   
</pre>

As such, we will close this issue now. If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by mark-casey** [comment]
Thanks. No worries on the delay. It clearly wasn't super widespread so (without discounting those of us who lost a little time finding it back when) it makes sense. I'd take me 15 minutes to even remember what I was doing when I hit this so I wouldn't be able to test it more than with the test case here, but knowing the two features can work together should the need arise is great.
